pkg/urbit/include/c/bile.h

```diff
@@ -0,0 +1,134 @@
+//! Use this function instead of using a direct field access so that any
+//! future changes to this module that change the binary file handle
+//! field representations or names don't require changes at call sites.
+//!
+//! @param[in] bil_u  Binary file handle. Must not be NULL.
+//!
+//! @return  Length of binary file in bytes.
+static inline c3_w
+c3_bile_len(c3_bile* const bil_u)
+{
+  return bil_u->len_w;
+}
+
+//! Get the path of a binary file.
+//!
+//! Use this function instead of using a direct field access so that any
+//! future changes to this module that change the binary file handle
+//! field representations or names don't require changes at call sites.
+//!
+//! @param[in] bil_u  Binary file handle. Must not be NULL.
+//!
+//! @return  Path of binary file.
+static inline c3_c*
+c3_bile_path(c3_bile* const bil_u)
+{
+  return bil_u->pax_c;
+}
+
+//! Open/create a binary file.
+//!
+//! @param[in] pax_c  Path to binary file.
+//!
+//! @return NULL  `pax_c` is NULL.
+//! @return NULL  File at `pax_c` could not be opened/created.
+//! @return NULL  File size of file at `pax_c` could not be determined.
+//! @return       Binary file handle.
+c3_bile*
+c3_bile_open(const c3_c* const pax_c);
+
+//! Append a raw byte array to a binary file.
+//!
+//! @param[in] bil_u  Binary file handle.
+//! @param[in] dat_v  Raw byte array. Can be NULL if `len_i` is 0.
+//! @param[in] len_i  Length of `dat_v` in bytes. If 0, clears the file.
+//!
+//! @return 0  Byte array could not be written to file.
+//! @return 1  Success.
+c3_t
+c3_bile_put_raw(c3_bile* const    bil_u,
+                const void* const dat_v,
+                const c3_w        len_w);
+
+//! Read a raw byte array from a binary file.
+//!
+//! @param[in] bil_u  Binary file handle.
+//! @param[in] dat_v  Destination address to write raw byte array to. Must have
+//!                   enough space for `c3_bile_len(bil_u)` bytes.
+//!
+//! @return 0  Byte array could not be read from file.
+//! @return 1  Success.
+c3_t
+c3_bile_get_raw(c3_bile* const bil_u, void* const dat_v);
+
+//! Write a string to `bil_u`.
+//!
+//! @param[in] bil_u  Binary file handle.
+//! @param[in] dat_c  String.
+//!
+//! @return 0  String could not be written to file.
+//! @return 1  Success.
+static inline c3_t
+c3_bile_put_str(c3_bile* const bil_u, const c3_c* const dat_c)
+{
+  return c3_bile_put_raw(bil_u, dat_c, 1 + strlen(dat_c));
+}
+
+//! Read a string from `bil_u`.
+//!
+//! @note This is functionally equivalent to c3_bile_get_raw() but can be used
+//!       to provide greater context at the call site.
+//!
+//! @param[in]  bil_u  Binary file handle.
+//! @param[out] dat_c  Destination address to write string to. Must have enough
+//!                    space for `c3_bile_len(bil_u)`  bytes.
+//!
+//! @return 0  String could not be read from file.
+//! @return 1  Success.
+static inline c3_t
+c3_bile_get_str(c3_bile* const bil_u, c3_c* const dat_c)
+{
+  return c3_bile_get_raw(bil_u, dat_c);
+}
```

Comment:
It'd be nice if we could ensure that the output is null-terminated here, but I think that would require a different api. In the meantime, can you add a comment noting that whether the string is null-terminated will depend on how the file was written?

Reply:
Done!